packages/contracts-bedrock/scripts/ChainAssertions.sol

```diff
@@ -0,0 +1,187 @@
+        if (startBlock != 0) {
+            require(config.startBlock() == startBlock);
+        } else {
+            require(config.startBlock() == block.number);
+        }
+    }
+
+    /// @notice Asserts that the legacy SystemConfig is setup correctly
+    function checkSystemConfigV0(address systemConfigProxy, DeployConfig cfg) internal view {
+        ISystemConfigV0 config = ISystemConfigV0(systemConfigProxy);
+        require(config.owner() == cfg.finalSystemOwner());
+        require(config.overhead() == cfg.gasPriceOracleOverhead());
+        require(config.scalar() == cfg.gasPriceOracleScalar());
+        require(config.batcherHash() == bytes32(uint256(uint160(cfg.batchSenderAddress()))));
+        require(config.unsafeBlockSigner() == cfg.p2pSequencerAddress());
+
+        ResourceMetering.ResourceConfig memory rconfig = Constants.DEFAULT_RESOURCE_CONFIG();
+        ResourceMetering.ResourceConfig memory resourceConfig = config.resourceConfig();
+        require(resourceConfig.maxResourceLimit == rconfig.maxResourceLimit);
+        require(resourceConfig.elasticityMultiplier == rconfig.elasticityMultiplier);
+        require(resourceConfig.baseFeeMaxChangeDenominator == rconfig.baseFeeMaxChangeDenominator);
+        require(resourceConfig.systemTxMaxGas == rconfig.systemTxMaxGas);
+        require(resourceConfig.minimumBaseFee == rconfig.minimumBaseFee);
+        require(resourceConfig.maximumBaseFee == rconfig.maximumBaseFee);
+    }
+
+    /// @notice Asserts that the L1CrossDomainMessenger is setup correctly
+    function checkL1CrossDomainMessenger(Types.ContractSet memory proxies, Vm vm) internal view {
+        L1CrossDomainMessenger messenger = L1CrossDomainMessenger(proxies.L1CrossDomainMessenger);
+        require(address(messenger.portal()) == proxies.OptimismPortal);
+        require(address(messenger.PORTAL()) == proxies.OptimismPortal);
+        bytes32 xdmSenderSlot = vm.load(address(messenger), bytes32(uint256(204)));
+        require(address(uint160(uint256(xdmSenderSlot))) == Constants.DEFAULT_L2_SENDER);
+    }
+
+    /// @notice Asserts that the L1StandardBridge is setup correctly
+    function checkL1StandardBridge(Types.ContractSet memory proxies, Vm vm) internal view {
+        L1StandardBridge bridge = L1StandardBridge(payable(proxies.L1StandardBridge));
+        require(address(bridge.MESSENGER()) == proxies.L1CrossDomainMessenger);
+        require(address(bridge.messenger()) == proxies.L1CrossDomainMessenger);
+        require(address(bridge.OTHER_BRIDGE()) == Predeploys.L2_STANDARD_BRIDGE);
+        require(address(bridge.otherBridge()) == Predeploys.L2_STANDARD_BRIDGE);
+        // Ensures that the legacy slot is modified correctly. This will fail
+        // during predeployment simulation on OP Mainnet if there is a bug.
+        bytes32 slot0 = vm.load(address(bridge), bytes32(uint256(0)));
+        require(slot0 == bytes32(uint256(Constants.INITIALIZER)));
```

Comment:
I am not sure the best way to do this in a portable way because sometimes the initialized value is at an offset. Modern openzeppelin puts it at a magic storage slot so we would never have this issue if we upgraded to v5.0.0. This check is nice to have here but just commenting on the fact that it cannot just work for all of the contracts because of the offset

Reply:
Yeah this wouldn't work for XDM contracts. What do you think about initializing `Constants.INITIALIZER` using some "random" starting value (ex: `keccak("ether's initializer")`), rather than 0 or 1, so we're more robust against false negatives during when asserting the initializer.